### ISSUE TITLE
ExternalRegistrySource reuses ExternalComponentRegistration instances, leading to use-after-dispose

### ISSUE THREAD

**Issue by sandersaares** [autofac/Autofac#960]
**Edit: updated issue description here. Original below and in comments.**

Consider a three-level lifetime scope hierarchy:

1. root
1. middle
1. child

Register type T in `root` with InstancePerLifetimeScope.

Register some irrelevant types in `middle` and in `child` (to materialize a scope-specific registry and not reuse the copy-on-write registry).

Now resolve type T from `middle`. This results in a new `ExternalComponentRegistration` being added to the registry of `middle`, based on the registration in `root`.

Now resolve type T from `child`. This results in **the same instance** of `ExternalComponentRegistration` being added to the registry of `child` as was added to `middle`.

Now dispose of `child`. The registration **still shared with `middle`** is disposed of!

Now attempt to resolve T from `middle`. You get ObjectDisposedException from the activator.

The root cause is that `ExternalRegistrySource.RegistrationsFor` will reuse existing instances from the parent if they are already ExternalComponentRegistrations.

Proposed fix: do not reuse instances of  `ExternalComponentRegistration`.

----------

I am trying to debug a registration disposal issue and request hints at what to look at, so I could properly analyze the situation and formulate a sensible bug report (or discover my error).

The situation is:

```
Exception thrown: 'System.ObjectDisposedException' in Autofac.dll
Instances cannot be created by this activator as it has already been disposed.
```

I have a root scope and a child scope. Neither is disposed. I am resolving from the child scope.

The failure appears to be when resolving my type T that has a parameter of type `ILifetimeScope` (used by my object T to create child scopes). The registration for `ILifetimeScope` has `IsDisposed=true` (looking in childScope.ComponentRegistry.Registrations).

In fact, registrations for all types inherited via `InstancePerLifetimeScope()` from my root scope are also `IsDisposed=true`. One registration that is defined directly on the child scope is `IsDisposed=false`.

Example of registering a type on the root scope (registration IsDisposed=true when exception occurs)

```
builder.RegisterType<DefaultSynchronousRetryMechanism>()
                .As<ISynchronousRetryMechanism>()
                .InstancePerLifetimeScope();
```

Example of creating child scope and registering a type on it (registration IsDisposed=false):

```
childScope = rootScope.BeginLifetimeScope(builder =>
{
    builder.RegisterType<JobExecutionAttempt>()
        .InstancePerLifetimeScope();
});
```

I later create a next layer child scope and try to resolve:

```
using (var jobScope = childScope.BeginLifetimeScope())
{
    var attempt = jobScope.Resolve<JobExecutionAttempt>();
    await attempt.TryExecuteAsync(nextJobId.Value);
}
```

`JobExecutionAttempt` takes parameter of type `ILifetimeScope` which triggers ObjectDisposedException in .Resolve().

I do not expect you to have a solution for me based on this meager info but I request input on what I should be looking at. It is very much unexpected to me that ILifetimeScope and inherited registrations are IsDisposed=true. What can cause registrations to be IsDisposed=true without the lifetime scope being disposed?

This seems to happen after some time - not on the first entry into the last using block but perhaps the second?

I will continue to dig and improve the issue report once I find more relevant data.

**Comment by sandersaares** [comment]
Okay here is what I see:

1. I have rootScope -> childScope -> jobScope -> actionScope layering.
1. A `using (jobScope)` block ends, disposing of jobScope (child of childScope).
1. This results in disposal of a `ScopeRestrictedRegistry` with 12 registrations. Two of these are registered in jobScope configuration. The other 10 are not.

I do not understand why the other 10 here are being disposed of (as they are registered in childScope not jobScope and the same instances later used again for other jobScopes).

**Comment by sandersaares** [comment]
Oh, I see it.

`ExternalRegistrySource.RegistrationsFor` has `r as ExternalComponentRegistration` as an alternative to creating a new registration. So if it sees that the thing it is registering is already an ExternalRegistration, it simply reuses it.

This means that my actionScope's registry reuses the `ExternalComponentRegistration` from its grandparent.

And then disposes of it! That's not right.

**Comment by chenmach** [comment]
I have the same problem after upgrading to 4.9.0 (was waiting for https://github.com/autofac/Autofac/issues/218 to use contextual loggers)

I was able to reproduce it using this sample :
https://gist.github.com/chenmach/fe9c029d84f2583dee6837741d5b2296

**Comment by sandersaares** [comment]
Looks like https://github.com/autofac/Autofac/pull/946 (included in 4.9.0) made it visible by throwing the exception, although the use-after-dispose was already happening (silently) in earlier versions.

**Comment by alexmg** [comment]
Thank you @sandersaares and @chenmach for the help. Looks like something that crept in under the radar a while ago.

I added a few more assertions to the unit test from @sandersaares and confirmed the issue fixes the scenario in the Gist from @chenmach.

**Comment by alexmg** [comment]
I have released `4.9.1` with this fix to NuGet.
